tests/rc_string_ut/rc_string_ut.c

```diff
@@ -26,40 +42,432 @@ static void on_umock_c_error(UMOCK_C_ERROR_CODE error_code)
+/* Tests_SRS_RC_STRING_01_017: [ rc_string_create_with_custom_free shall succeed and return a non-NULL handle. ]*/
+TEST_FUNCTION(rc_string_create_with_custom_free_succeeds)
+{
+    // arrange
+    const char const_test_string[] = "goguletz";
+    char* test_string = (char*)test_malloc_func(sizeof(const_test_string));
+    ASSERT_IS_NOT_NULL(test_string);
+
+    (void)memcpy(test_string, const_test_string, sizeof(const_test_string));
+    umock_c_reset_all_calls();
+
+    STRICT_EXPECTED_CALL(malloc(IGNORED_ARG));
+
+    // act
+    THANDLE(RC_STRING) rc_string = rc_string_create_with_custom_free(test_string, test_free_func, test_string);
+
+    // assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    ASSERT_ARE_EQUAL(void_ptr, test_string, rc_string->string);
+    ASSERT_IS_NOT_NULL(rc_string);
+
+    // cleanup
+    THANDLE_DEC_REF(RC_STRING)(rc_string);
+}
+
+/* Tests_SRS_RC_STRING_01_015: [ rc_string_create_with_custom_free shall allocate memory for the THANDLE(RC_STRING). ]*/
+/* Tests_SRS_RC_STRING_01_016: [ rc_string_create_with_custom_free shall associate string, free_func and free_func_context with the new handle. ]*/
+/* Tests_SRS_RC_STRING_01_017: [ rc_string_create_with_custom_free shall succeed and return a non-NULL handle. ]*/
+TEST_FUNCTION(rc_string_create_with_custom_free_with_empty_string_succeeds)
+{
+    // arrange
+    const char const_test_string[] = "";
+    char* test_string = (char*)test_malloc_func(sizeof(const_test_string));
+    ASSERT_IS_NOT_NULL(test_string);
+
+    (void)memcpy(test_string, const_test_string, sizeof(const_test_string));
+    umock_c_reset_all_calls();
+
+    STRICT_EXPECTED_CALL(malloc(IGNORED_ARG));
+
+    // act
+    THANDLE(RC_STRING) rc_string = rc_string_create_with_custom_free(test_string, test_free_func, test_string);
+
+    // assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    ASSERT_ARE_EQUAL(void_ptr, test_string, rc_string->string);
+    ASSERT_IS_NOT_NULL(rc_string);
+
+    // cleanup
+    THANDLE_DEC_REF(RC_STRING)(rc_string);
+}
+
+/* Tests_SRS_RC_STRING_01_014: [ free_func_context shall be allowed to be NULL. ]*/
+TEST_FUNCTION(rc_string_create_with_custom_free_with_do_nothing_free_and_NULL_context_succeeds)
+{
+    // arrange
+    const char const_test_string[] = "";
+    char* test_string = (char*)test_malloc_func(sizeof(const_test_string));
+    ASSERT_IS_NOT_NULL(test_string);
+
+    (void)memcpy(test_string, const_test_string, sizeof(const_test_string));
+    umock_c_reset_all_calls();
+
+    STRICT_EXPECTED_CALL(malloc(IGNORED_ARG));
+
+    // act
+    THANDLE(RC_STRING) rc_string = rc_string_create_with_custom_free(test_string, test_free_func_do_nothing, NULL);
+
+    // assert
+    ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+    ASSERT_ARE_EQUAL(void_ptr, test_string, rc_string->string);
+    ASSERT_IS_NOT_NULL(rc_string);
```

Comment:
>ASSERT_IS_NOT_NULL(rc_string); [](start = 4, length = 30)

switch it with the line above. here and everywhere, I fear this comment might come late now :)

Reply:
switched.

---
In reply to: [460263108](https://github.com/Azure/azure-c-util/pull/52#discussion_r460263108) [](ancestors = 460263108)